File: functions/test_prompt.py
import datetime
import json
from dotenv import load_dotenv
import numpy as np
import torch
from angle_emb import AnglE
from openai import OpenAI
from data_classes import CaseScore, Prompt, Context, Bio, TestCase, TestResults

# Load the environment variables and set constants
load_dotenv()
openai_client = OpenAI()
OPENAI_MODEL = "gpt-3.5-turbo"
TEMPERATURE = 0.7
MAX_TOKENS = 50
NUM_RESPONSES = 4
EMBEDDING_MODEL_NAME = 'WhereIsAI/UAE-Large-V1'


def embed_texts(texts: list[str]) -> np.ndarray:
    # Check if the model exists
    try:
        angle = AnglE.from_pretrained(
            EMBEDDING_MODEL_NAME, pooling_strategy='cls')
    except Exception as e:
        raise ValueError(
            f"Failed to load the model '{EMBEDDING_MODEL_NAME}'. Error: {str(e)}")

    # Check if CUDA is available, if not use CPU
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    angle = angle.to(device)

    embeddings = angle.encode(texts, to_numpy=True)
    # np.save("./embs.npy", embeddings)  # TODO: Not sure what we want to do here
    return embeddings


def cosine_similarity(embed1: np.ndarray, embed2: np.ndarray) -> float:
    '''
        Compares two embedding vectors and returns the cosine similarity.
        Output range is from -1 to 1, where 1 means the vectors are identical.
    '''
    return np.dot(embed1, embed2) / (np.linalg.norm(embed1) * np.linalg.norm(embed2))


def get_gpt_completion(prompt: str) -> list[str]:
    '''Makes a request to the OpenAI API and returns the response'''
    chat_completion = openai_client.chat.completions.create(
        messages=[
            {
                "role": "user",
                "content": prompt,
            }
        ],
        model=OPENAI_MODEL,
        max_tokens=MAX_TOKENS,
        temperature=TEMPERATURE,
        n=NUM_RESPONSES
    )

    return [choice.message.content for choice in chat_completion.choices]


def test_prompt(prompt: Prompt, test_cases: list[TestCase]) -> TestResults:
    '''Executes each test case against the given prompt and prints the score'''
    scores: list[CaseScore] = []
    for test_case in test_cases:
        gpt_completions = get_gpt_completion(prompt.prompt.format(
            utterance=test_case.utterance,
            setting=test_case.context.setting,
            tone=test_case.context.tone,
            conversation_type=test_case.context.conversationType,
            name=test_case.bio.name,
            age=test_case.bio.age,
            about_me=test_case.bio.aboutMe
        ))
        print(f"Test case {test_case.id} against prompt {prompt.id}")
        print(f"GPT completions: {gpt_completions}")

        # Get the embeddings for the GPT responses and the good responses
        gpt_embeddings = embed_texts(gpt_completions)
        good_embeddings = embed_texts(test_case.goodCompletions)

        # Average the embeddings
        avg_gpt_embedding = np.mean(gpt_embeddings, axis=0)
        avg_good_embedding = np.mean(good_embeddings, axis=0)

        # Compare the embeddings and store the score
        cosine_similarity_score = cosine_similarity(
            avg_gpt_embedding, avg_good_embedding)
        scores.append(
            CaseScore(test_case.id, cosine_similarity_score))

    avg_cosine_similarity_score = np.mean(
        [score.cosine_similarity_score for score in scores])
    print(f"Average cosine similarity score: {avg_cosine_similarity_score}")

    return TestResults(
        prompt.id,
        OPENAI_MODEL,
        EMBEDDING_MODEL_NAME,
        TEMPERATURE,
        MAX_TOKENS,
        NUM_RESPONSES,
        scores,
        avg_cosine_similarity_score
    )

File: functions/src/testPrompt.ts
import * as logger from "firebase-functions/logger";
import {onRequest} from "firebase-functions/v2/https";

import {
  PromptCandidate,
  TestCase,
  TestCaseResult,
  TestResultsStatus,
} from "../../shared/types";
import {getUnixTimestamp} from "../../shared/utils";
import {PRODUCTION_APP_URL} from "./constants";
import {
  getPromptById,
  getPromptTestResultsRef,
  initializePromptTestResultsRecord,
  readTestCases,
} from "./firebaseUtils";
import {runPromptTestCase} from "./runPromptTestCase";

const OPENAI_MODEL = "gpt-3.5-turbo";
const TEMPERATURE = 0.7;
const MAX_TOKENS = 50;
const NUM_RESPONSES = 4;
const EMBEDDING_MODEL_NAME = "WhereIsAI/UAE-Large-V1";

/**
 * Run all test cases in the DB against the provided prompt.
 * Persist the results in the DB.
 */
export const testPromptHandler = onRequest(
  {cors: PRODUCTION_APP_URL},
  async (req, res) => {
    const data = req.body;
    const promptId = data.promptId;
    if (!promptId) {
      res.status(400).send("No promptId parameter provided");
      return;
    }
    const prompt: PromptCandidate | null = await getPromptById(promptId);
    if (!prompt) {
      res.status(400).send(`Prompt with id ${promptId} not found`);
      return;
    }

    const testCases: TestCase[] = await readTestCases();

    const testResultsId = await initializePromptTestResultsRecord(
      testCases,
      promptId,
      OPENAI_MODEL,
      EMBEDDING_MODEL_NAME,
      TEMPERATURE,
      MAX_TOKENS,
      NUM_RESPONSES
    );

    for (const testCase of testCases) {
      if (!testCase.id) {
        logger.error("Test case has no id");
        continue;
      }

      logger.info(
        `Testing prompt ${promptId} against test case ${testCase.id}`
      );

      const {llmCompletions, cosineSimilarityScore} = await runPromptTestCase(
        prompt,
        testCase,
        OPENAI_MODEL,
        EMBEDDING_MODEL_NAME,
        TEMPERATURE,
        MAX_TOKENS,
        NUM_RESPONSES
      );
      logger.info(
        `Savings results for case ${testCase.id} against prompt ${promptId}:
       cosineSimilarityScore = ${cosineSimilarityScore}`
      );
      await saveTestCaseResult(
        testResultsId,
        testCase.id,
        cosineSimilarityScore,
        llmCompletions
      );
    }

    console.log("All tests completed");
    res.send({message: "Tests completed"});
  }
);

/**
 * Save the results of a test case to the DB.
 * @param {string} promptTestResultsId The ID of the prompt test results record.
 * @param {string} testCaseId The ID of the test case.
 * @param {number} cosineSimilarityScore The cosine similarity score.
 * @param {string[]} llmCompletions The completions from the LLM.
 */
async function saveTestCaseResult(
  promptTestResultsId: string,
  testCaseId: string,
  cosineSimilarityScore: number,
  llmCompletions: string[]
) {
  const testCaseResult: TestCaseResult = {
    testCaseId: testCaseId,
    status: TestResultsStatus.COMPLETED,
    cosineSimilarityScore: cosineSimilarityScore,
    llmCompletions: llmCompletions,
    dateCompletedUtc: getUnixTimestamp(),
  };
  const resultRef = getPromptTestResultsRef().child(promptTestResultsId);
  await resultRef
    .child("testCaseResults")
    .child(testCaseId)
    .set(testCaseResult);
}
